File: dsa-problems/Tries/solutions/javascript/TRI-007-minimum-unique-prefix-lengths.js
const readline = require("readline");

class TrieNode {
  constructor() {
    this.children = new Map();
    this.count = 0;
    this.isEnd = false;
  }
}

class Solution {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word) {
    let node = this.root;
    for (const char of word) {
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode());
      }
      node = node.children.get(char);
      node.count++;
    }
    node.isEnd = true;
  }

  findMinLength(word) {
    let node = this.root;
    for (let i = 0; i < word.length; i++) {
      node = node.children.get(word[i]);
      if (node.count === 1) {
        return i + 1;
      }
    }
    return word.length;
  }

  findMinimumPrefixLengths(words) {
    // Build trie with counts
    for (const word of words) {
      this.insert(word);
    }

    // Find minimum prefix length for each word
    const result = [];
    for (const word of words) {
      result.push(this.findMinLength(word));
    }

    return result;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const lines = [];
rl.on("line", (line) => {
  lines.push(line);
}).on("close", () => {
  const n = parseInt(lines[0]);
  const words = [];
  for (let i = 1; i <= n; i++) {
    words.push(lines[i].trim());
  }

  const solution = new Solution();
  const result = solution.findMinimumPrefixLengths(words);

  console.log("[" + result.join(",") + "]");
});

File: dsa-problems/Tries/solutions/cpp/TRI-007-minimum-unique-prefix-lengths.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    int count = 0;
    bool isEnd = false;
};

class Solution {
private:
    TrieNode* root;

    void insert(const string& word) {
        TrieNode* node = root;
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
            node->count++;
        }
        node->isEnd = true;
    }

    int findMinLength(const string& word) {
        TrieNode* node = root;
        for (int i = 0; i < word.length(); i++) {
            node = node->children[word[i]];
            if (node->count == 1) {
                return i + 1;
            }
        }
        return word.length();
    }

public:
    Solution() {
        root = new TrieNode();
    }

    vector<int> findMinimumPrefixLengths(vector<string>& words) {
        // Build trie with counts
        for (const string& word : words) {
            insert(word);
        }

        // Find minimum prefix length for each word
        vector<int> result;
        for (const string& word : words) {
            result.push_back(findMinLength(word));
        }

        return result;
    }
};

int main() {
    int n;
    cin >> n;
    cin.ignore();

    vector<string> words(n);
    for (int i = 0; i < n; i++) {
        getline(cin, words[i]);
    }

    Solution solution;
    vector<int> result = solution.findMinimumPrefixLengths(words);

    cout << "[";
    for (int i = 0; i < result.size(); i++) {
        cout << result[i];
        if (i < result.size() - 1) cout << ",";
    }
    cout << "]" << endl;

    return 0;
}
